Sample: Trace utility network

shubham7109 opened the pull request:
## Description
PR to add a new Kotlin sample _"Trace utility network"_ in `Utility Networks` category.

## Links and Data
Sample issue: [#2935](https://devtopia.esri.com/runtime/kotlin/issues/2935) 
- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/) Job for this PR has been run
  - [x] link: https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/68/

## What To Review

- This is a fairly SDK intricate sample. Review the view model code and look for simplifications/code optimizations. 
- Compose UI using a bottom sheet auto disappears while interacting with the map. Does the UI design seem straightforward and intuitive? 
-  Any sample suggestions/enhancements. 

## How to Test

- Run the sample and try performing various traces on junctions and edges.
- Try using multiple starting points. 
- Verify if the MapView results are as expected.

sorenoid reviewed (changes requested):
Just a few minor suggestions. I am running it now but wanted to give this feedback asap.

sorenoid commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/screens/TraceUtilityNetworkScreen.kt:
```diff
@@ -90,6 +90,9 @@ fun TraceUtilityNetworkScreen(sampleName: String) {
     val mapViewModel: TraceUtilityNetworkViewModel = viewModel()
 
+    // Load map, utility network, and adds layers. 
+    LaunchedEffect(Unit) { mapViewModel.initializeTrace() }
```
sorry to string you along. This will run every time the composition recomposes, but the viewmodel only needs to initialize once. Is there a way to check if the viewmodel has been initialized?

shubham7109 replied:
Good point. Added a check in the `LaunchedEffect` to run if the network is not loaded.

shubham7109 commented:
@sorenoid I have hopefully fixed the UI issue, I was only able to test it through Compose previews. I added the feedback to check if the ViewModel is initialized. 

Ready for your re-review.

sorenoid reviewed (approved):
👍

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/screens/TraceUtilityNetworkScreen.kt:
```diff
@@ -0,0 +1,459 @@
+        if (selectedTraceType == null) {
+            focusManager.clearFocus()
+            selectedTraceName = "Select a trace type"
+        } else selectedTraceName = selectedTraceType.javaClass.simpleName
```
kind of confusing to me having an if with braces and the else without

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/MainActivity.kt:
```diff
@@ -0,0 +1,62 @@
+        // authentication with an API key or named user is
+        // required to access basemaps and other location services
+        ArcGISEnvironment.apiKey = ApiKey.create(BuildConfig.ACCESS_TOKEN)
+        ArcGISEnvironment.applicationContext = application.applicationContext
```
Is this needed? Runs fine for me when I comment it out. If it is needed can we add a comment as to why

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,621 @@
+    val mapViewProxy = MapViewProxy()
+
+    // The utility network used for tracing.
+    private val utilityNetwork: UtilityNetwork
```
What's the advantage of setting a get() instead of assigning the value with an = ?

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,621 @@
+        ).onSuccess { tokenCredential ->
+            // Add the loaded token credential to the app session's authenticationManager
+            ArcGISEnvironment.authenticationManager.arcGISCredentialStore.add(tokenCredential)
+            // Load the NapervilleElectric web-map
```
```suggestion
            // Load the Naperville electric web-map
```

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,621 @@
+        // Find the code matching the asset group name in the feature's attributes
+        val assetGroupCode = identifiedFeature.attributes["assetgroup"] as Int
+        // Find the network source's asset group with the matching code
+        utilityNetworkSource.assetGroups.filter { it.code == assetGroupCode }[0].assetTypes
```
```suggestion
        utilityNetworkSource.assetGroups.first { it.code == assetGroupCode }.assetTypes
```

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,621 @@
+        // Find the network source's asset group with the matching code
+        utilityNetworkSource.assetGroups.filter { it.code == assetGroupCode }[0].assetTypes
+            // Find the asset group type code matching the feature's asset type code
+            .filter { it.code == identifiedFeature.attributes["assettype"].toString().toInt() }[0]
```
```suggestion
            .first { it.code == identifiedFeature.attributes["assettype"].toString().toInt() }
```

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,621 @@
+                }
+
+            // Get the utility trace result's first result as a utility element trace result
+            (traceResults[0] as? UtilityElementTraceResult)?.let { utilityElementTraceResult ->
```
```suggestion
            (traceResults.first() as? UtilityElementTraceResult)?.let { utilityElementTraceResult ->
```

TADraeseke reviewed (changes requested):
Wow @shubham7109 -- epic sample! Just some thoughts below!

shubham7109 commented:
@sorenoid @TADraeseke I tweaked the sample workflow to enable a default `Connected` trace type and a default `Start` point type. This way, the simplest workflow is to tap on map -> then press the Trace button. This helped clean up some of the null safety codes. 



I added the feedback changes, ready for your review!

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,617 @@
+
+    // The utility network used for tracing.
+    private val utilityNetwork: UtilityNetwork
+        get() = arcGISMap.utilityNetworks[0]
```
```suggestion
        get() = arcGISMap.utilityNetworks.first()
```

TADraeseke commented on samples/trace-utility-network/src/main/java/com/esri/arcgismaps/sample/traceutilitynetwork/components/TraceUtilityNetworkViewModel.kt:
```diff
@@ -0,0 +1,617 @@
+                        // Create a utility element
+                        utilityNetwork.createElementOrNull(
+                            arcGISFeature = identifiedFeature,
+                            terminal = terminals[0]
```
```suggestion
                            terminal = terminals.first()
```

TADraeseke reviewed (approved):
LGTM! Just a couple more optional .first()s below.

shubham7109 commented:
@TADraeseke I added a few changes, mainly to enhance the sample's experience. Previously, an error dialog would be shown if the trace results returned an empty list. This could be done by adding a starting point between two barriers. 


![trace_image](https://github.com/user-attachments/assets/83425ac7-3f23-4629-8d82-69d27ef843ea)




And, I added some minor UI changes regarding errors and the fraction decimals shown. Would appreciate it if you could review these changes.

TADraeseke reviewed (approved):
LGTM!
